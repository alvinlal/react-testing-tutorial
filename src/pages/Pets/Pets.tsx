import { useEffect, useState } from 'react';
import Cards from './components/Cards/Cards';
import Filter, { FilterOptions } from './components/Filter/Filter';
import './Pets.css';

const Pets: React.FC = () => {
  const [cats, setCats] = useState<Cat[]>([]);
  const [filteredCats, setFilteredCats] = useState<Cat[]>([]);
  const [filters, setFilters] = useState<FilterOptions>({
    gender: 'any',
    favorite: 'any',
  });
  const [status, setStatus] = useState('idle');

  const fetchCats = async () => {
    try {
      setStatus('loading');
      const data = await fetch('https://631c5fa04fa7d3264caca918.mockapi.io/api/cats').then(res => {
        if (!res.ok) {
          throw new Error('something went wrong');
        }
        return res.json();
      });
      if (!data.error) {
        setCats(data);
        setFilteredCats(data);
        setStatus('success');
      }
    } catch (err) {
      setStatus('error');
    }
  };

  const updateFavoured = (index: number, favoured: boolean) => {
    const updatedCats = [...cats];
    updatedCats[index].favoured = favoured;
    setCats(updatedCats);
  };

  useEffect(() => {
    fetchCats();
  }, []);

  useEffect(() => {
    let catsFiltered = [...cats];
    if (filters.gender !== 'any') {
      catsFiltered = catsFiltered.filter(cat => cat.gender === filters.gender);
    }
    if (filters.favorite !== 'any') {
      catsFiltered = catsFiltered.filter(
        cat => cat.favoured === (filters.favorite === 'favoured' ? true : false)
      );
    }
    setFilteredCats(catsFiltered);
    // eslint-disable-next-line
  }, [filters]);

  return (
    <div className='container'>
      <div className='app-container'>
        <Filter setFilters={setFilters} />
        {status === 'loading' && <p>loading cats...</p>}
        {status === 'error' && <h3>something went wrong</h3>}
        {status === 'success' && <Cards cats={filteredCats} updateFavoured={updateFavoured} />}
      </div>
    </div>
  );
};

export default Pets;
