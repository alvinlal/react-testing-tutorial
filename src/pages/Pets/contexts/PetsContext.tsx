import React, { useState, useEffect } from 'react';

const PetsContext = React.createContext<{
  cats: Cat[];
  setCats: React.Dispatch<React.SetStateAction<Cat[]>>;
  updateFavoured: (index: number, favoured: boolean) => void;
  status: string;
}>({ cats: [], setCats: () => {}, updateFavoured: () => {}, status: 'idle' });

const PetsProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [cats, setCats] = useState<Cat[]>([]);
  const [status, setStatus] = useState('idle');

  const updateFavoured = (index: number, favoured: boolean) => {
    const updatedCats = [...cats];
    updatedCats[index].favoured = favoured;
    setCats(updatedCats);
  };

  const fetchCats = async () => {
    try {
      setStatus('loading');
      const data = await fetch('https://631c5fa04fa7d3264caca918.mockapi.io/api/cats').then(res => {
        if (!res.ok) {
          throw new Error('something went wrong');
        }
        return res.json();
      });
      if (!data.error) {
        setCats(data);
        setStatus('success');
      }
    } catch (err) {
      setStatus('error');
    }
  };

  useEffect(() => {
    fetchCats();
  }, []);

  return (
    <PetsContext.Provider value={{ cats, setCats, updateFavoured, status }}>
      {children}
    </PetsContext.Provider>
  );
};

export { PetsContext, PetsProvider };
