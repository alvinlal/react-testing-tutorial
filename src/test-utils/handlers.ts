import { rest } from 'msw';
import cats from './mocks/cats.json';

const handlers = [
  rest.get('https://631c5fa04fa7d3264caca918.mockapi.io/api/cats', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(cats));
  }),
];

export default handlers;
