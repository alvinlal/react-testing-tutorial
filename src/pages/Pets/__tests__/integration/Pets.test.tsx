/**
 * Requirements
 * 1. Should show 5 Card component if the api request is successfull
 * 2. Should show loader when the api request is loading
 * 3. Should show error message when the api request has failed
 * 4. Should be able to filter by male cats
 * 5. Should be able to filter by female cats
 * 5. Should be able to filter by favorited cats
 * 6. Should be able to filter by not favorited cats
 * 7. Should be able to filter by favoured male cats
 */

import { render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { server, rest } from '../../../../test-utils/server';
import { PetsProvider } from '../../contexts/PetsContext';
import Pets from '../../Pets';

describe('Pets.tsx', () => {
  it('Should show 5 Card component if the api request is successfull', async () => {
    render(
      <PetsProvider>
        <Pets />
      </PetsProvider>
    );
    const cards = await screen.findAllByTestId('cat-card');
    expect(cards.length).toBe(5);
  });

  it('Should show loader when the api request is loading', async () => {
    render(
      <PetsProvider>
        <Pets />
      </PetsProvider>
    );
    const loader = await screen.findByText('loading cats...');
    expect(loader).toBeInTheDocument();
  });

  it('Should show error message when the api request has failed', async () => {
    server.use(
      rest.get('https://631c5fa04fa7d3264caca918.mockapi.io/api/cats', async (req, res, ctx) => {
        return res(ctx.status(500), ctx.json({ error: 'internal server error' }));
      })
    );

    render(
      <PetsProvider>
        <Pets />
      </PetsProvider>
    );

    expect(
      await screen.findByRole('heading', {
        name: 'something went wrong',
      })
    ).toBeInTheDocument();
  });

  it('Should be able to filter by male cats', async () => {
    render(
      <PetsProvider>
        <Pets />
      </PetsProvider>
    );
    const cards = await screen.findAllByTestId('cat-card');
    const genderFilter = screen.getByLabelText('Gender');
    userEvent.selectOptions(genderFilter, 'male');
    const maleCatCards = screen.getAllByTestId('cat-card');
    expect(maleCatCards).toStrictEqual([cards[1], cards[3]]);
  });

  it('Should be able to filter by female cats', async () => {
    render(
      <PetsProvider>
        <Pets />
      </PetsProvider>
    );
    const cards = await screen.findAllByTestId('cat-card');
    const genderFilter = screen.getByLabelText('Gender');
    userEvent.selectOptions(genderFilter, 'female');
    const maleCatCards = screen.getAllByTestId('cat-card');
    expect(maleCatCards).toStrictEqual([cards[0], cards[2], cards[4]]);
  });

  it('Should be able to filter by favorited cats', async () => {
    render(
      <PetsProvider>
        <Pets />
      </PetsProvider>
    );
    const cards = await screen.findAllByTestId('cat-card');
    userEvent.click(within(cards[0]).getByRole('button'));
    userEvent.click(within(cards[3]).getByRole('button'));
    const favoriteFilter = screen.getByLabelText('Favorite');
    userEvent.selectOptions(favoriteFilter, 'favoured');
    const favoritedCatCards = screen.getAllByTestId('cat-card');
    expect(favoritedCatCards).toStrictEqual([cards[0], cards[3]]);
  });

  it('Should be able to filter by not favorited cats', async () => {
    render(
      <PetsProvider>
        <Pets />
      </PetsProvider>
    );
    const cards = await screen.findAllByTestId('cat-card');
    userEvent.click(within(cards[0]).getByRole('button'));
    userEvent.click(within(cards[4]).getByRole('button'));
    const favoriteFilter = screen.getByLabelText('Favorite');
    userEvent.selectOptions(favoriteFilter, 'not favoured');
    const notFavoritedCatCards = screen.getAllByTestId('cat-card');
    expect(notFavoritedCatCards).toStrictEqual([cards[1], cards[2], cards[3]]);
  });

  it('Should be able to filter by favoured male cats', async () => {
    render(
      <PetsProvider>
        <Pets />
      </PetsProvider>
    );
    const cards = await screen.findAllByTestId('cat-card');
    userEvent.click(within(cards[0]).getByRole('button'));
    userEvent.click(within(cards[3]).getByRole('button'));
    const favoriteFilter = screen.getByLabelText('Favorite');
    userEvent.selectOptions(favoriteFilter, 'favoured');
    const genderFilter = screen.getByLabelText('Gender');
    userEvent.selectOptions(genderFilter, 'male');
    const favouredMaleCatCards = screen.getAllByTestId('cat-card');
    expect(favouredMaleCatCards).toStrictEqual([cards[3]]);
  });
});
