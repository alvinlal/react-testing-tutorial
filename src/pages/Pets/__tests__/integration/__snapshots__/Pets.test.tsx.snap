// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pets.tsx Should match api request error state snapshot 1`] = `
<div
  className="container"
>
  <div
    className="app-container"
  >
    <div
      className="pet-filter-container"
    >
      <div
        className="filter-container"
      >
        <label
          htmlFor="favorite"
        >
          Favorite
        </label>
        <select
          className="form-select"
          id="favorite"
          name="favorite"
          onChange={[Function]}
        >
          <option
            value="any"
          >
            any
          </option>
          <option
            value="favoured"
          >
            favoured
          </option>
          <option
            value="not favoured"
          >
            not favoured
          </option>
        </select>
      </div>
      <div
        className="filter-container"
      >
        <label
          htmlFor="gender"
        >
          Gender
        </label>
        <select
          className="form-select"
          id="gender"
          name="gender"
          onChange={[Function]}
        >
          <option
            value="any"
          >
            any
          </option>
          <option
            value="male"
          >
            male
          </option>
          <option
            value="female"
          >
            female
          </option>
        </select>
      </div>
    </div>
    <h3>
      something went wrong
    </h3>
  </div>
</div>
`;

exports[`Pets.tsx Should match api request loading state snapshot 1`] = `
<div
  className="container"
>
  <div
    className="app-container"
  >
    <div
      className="pet-filter-container"
    >
      <div
        className="filter-container"
      >
        <label
          htmlFor="favorite"
        >
          Favorite
        </label>
        <select
          className="form-select"
          id="favorite"
          name="favorite"
          onChange={[Function]}
        >
          <option
            value="any"
          >
            any
          </option>
          <option
            value="favoured"
          >
            favoured
          </option>
          <option
            value="not favoured"
          >
            not favoured
          </option>
        </select>
      </div>
      <div
        className="filter-container"
      >
        <label
          htmlFor="gender"
        >
          Gender
        </label>
        <select
          className="form-select"
          id="gender"
          name="gender"
          onChange={[Function]}
        >
          <option
            value="any"
          >
            any
          </option>
          <option
            value="male"
          >
            male
          </option>
          <option
            value="female"
          >
            female
          </option>
        </select>
      </div>
    </div>
    <p>
      loading cats...
    </p>
  </div>
</div>
`;

exports[`Pets.tsx Should match api request success state snapshot 1`] = `
<div
  className="container"
>
  <div
    className="app-container"
  >
    <div
      className="pet-filter-container"
    >
      <div
        className="filter-container"
      >
        <label
          htmlFor="favorite"
        >
          Favorite
        </label>
        <select
          className="form-select"
          id="favorite"
          name="favorite"
          onChange={[Function]}
        >
          <option
            value="any"
          >
            any
          </option>
          <option
            value="favoured"
          >
            favoured
          </option>
          <option
            value="not favoured"
          >
            not favoured
          </option>
        </select>
      </div>
      <div
        className="filter-container"
      >
        <label
          htmlFor="gender"
        >
          Gender
        </label>
        <select
          className="form-select"
          id="gender"
          name="gender"
          onChange={[Function]}
        >
          <option
            value="any"
          >
            any
          </option>
          <option
            value="male"
          >
            male
          </option>
          <option
            value="female"
          >
            female
          </option>
        </select>
      </div>
    </div>
    <p>
      loading cats...
    </p>
  </div>
</div>
`;
